File: frontend/src/services/deviceService.ts
/**
 * Device service — CRUD for Desktop, Laptop, and Tablet assets.
 *
 * Endpoints (Javalin backend):
 *   GET    /api/devices        → AnyDevice[]  (chapter filtering is enforced via JWT claims)
 *   GET    /api/devices/:id    → AnyDevice
 *   POST   /api/devices        → AnyDevice   (body: AnyDevice)
 *   PUT    /api/devices/:id    → AnyDevice   (body: AnyDevice)
 */

import { apiGet, apiGetOrNull, apiPost, apiPutVoid, buildQuery } from "./api";
import type {
  AnyDevice,
  IdResponse,
  InsertDesktopRequest,
  InsertLaptopRequest,
  InsertTabletRequest,
  AvgTimeInInventoryResponse,
  CompletionRateResponse,
  ChapterActivityStatsResponse,
  ChapterInventorySummary,
  DashboardCountsResponse,
  MonthlyCountPoint,
  MonthlyValuePoint,
} from "../types/inventory";
import type { DeviceChangelogEntry } from "../types/changelog";
import type { DeviceChangelogResponse } from "../types/inventory";
import { getChapters } from "./chapterService";

export interface DeviceListParams {
  pageKey: number;
  pageSize: number;
  search?: string;
  /** "Desktop" | "Laptop" | "Tablet" */
  type?: string;
  /** Exact device status, e.g. "Not Started". */
  status?: string;
  /** Chapter id to restrict to (within the user's access). */
  chapter?: number;
  includeDonated?: boolean;
  includeScrapped?: boolean;
  donorId?: number;
  recipientId?: number;
  /** Sort key matching DeviceList columns, e.g. "manufacturer". */
  sort?: string;
  dir?: "asc" | "desc";
}

export async function getDevices(params: DeviceListParams): Promise<AnyDevice[]> {
  return apiGet<AnyDevice[]>(`/devices${buildQuery({ ...params })}`);
}

export async function getChapterInventorySummary(): Promise<ChapterInventorySummary[]> {
  return apiGet<ChapterInventorySummary[]>("/devices/stats/chapter-inventory");
}

/** Returns null when no device with the given ID exists. */
export async function getDevice(id: number): Promise<AnyDevice | null> {
  return apiGetOrNull<AnyDevice>(`/devices/${id}`);
}

export async function createDevice(device: AnyDevice): Promise<AnyDevice> {
  const chapters = await getChapters();
  const chapter = chapters.find((c) => c.name === device.chapter);
  if (!chapter) throw new Error(`Unknown chapter: ${device.chapter}`);

  const chapterId = chapter.id;
  const assetId = device.id > 0 ? device.id : undefined;
  const common = {
    chapterId,
    assetId,
    manufacturer: device.manufacturer,
    model: device.model,
    year: device.year,
    status: device.status,
    cpu: device.cpu ?? undefined,
    ram: device.ram || undefined,
    ramGeneration: device.ramGeneration ?? undefined,
    storageAmount: device.storage || undefined,
    storageType: device.storageType ?? undefined,
    value: device.value ?? undefined,
    acquisitionDate: device.acquisitionDate ?? undefined,
    operatingSystem: device.operatingSystem ?? undefined,
  };

  let newId: number;
  if (device.type === "Desktop") {
    const body: InsertDesktopRequest = {
      ...common,
      hasWifi: device.hasWifi ?? undefined,
    };
    newId = (await apiPost<IdResponse>("/devices/desktop", body)).id;
  } else if (device.type === "Laptop") {
    const body: InsertLaptopRequest = {
      ...common,
      includesCharger: device.includesCharger ?? undefined,
      designBatteryCapacity: device.designBatteryCapacity ?? undefined,
      actualBatteryCapacity: device.actualBatteryCapacity ?? undefined,
    };
    newId = (await apiPost<IdResponse>("/devices/laptop", body)).id;
  } else if (device.type === "Tablet") {
    const body: InsertTabletRequest = {
      ...common,
      includesCharger: device.includesCharger ?? undefined,
      workingBattery: device.workingBattery ?? undefined,
    };
    newId = (await apiPost<IdResponse>("/devices/tablet", body)).id;
  } else {
    throw new TypeError("Unrecognized device type");
  }

  // Backend returns 201 with the new asset id; fetch the full record by id.
  return apiGet<AnyDevice>(`/devices/${newId}`);
}

export async function updateDevice(id: number, updates: AnyDevice): Promise<AnyDevice> {
  const chapters = await getChapters();
  const chapter = chapters.find((c) => c.name === updates.chapter);
  if (!chapter) throw new Error(`Unknown chapter: ${updates.chapter}`);

  const common = {
    chapterId: chapter.id,
    assetId: id,
    manufacturer: updates.manufacturer,
    model: updates.model,
    year: updates.year,
    status: updates.status,
    cpu: updates.cpu ?? undefined,
    ram: updates.ram || undefined,
    ramGeneration: updates.ramGeneration ?? undefined,
    storageAmount: updates.storage || undefined,
    storageType: updates.storageType ?? undefined,
    value: updates.value ?? undefined,
    acquisitionDate: updates.acquisitionDate ?? undefined,
    operatingSystem: updates.operatingSystem ?? undefined,
    donorId: updates.donorId ?? undefined,
    recipientId: updates.recipientId ?? undefined,
  };

  if (updates.type === "Desktop") {
    const body: InsertDesktopRequest = { ...common, hasWifi: updates.hasWifi ?? undefined };
    await apiPutVoid(`/devices/desktop/${id}`, body);
  } else if (updates.type === "Laptop") {
    const body: InsertLaptopRequest = {
      ...common,
      includesCharger: updates.includesCharger ?? "Unknown",
      designBatteryCapacity: updates.designBatteryCapacity ?? undefined,
      actualBatteryCapacity: updates.actualBatteryCapacity ?? undefined,
    };
    await apiPutVoid(`/devices/laptop/${id}`, body);
  } else if (updates.type === "Tablet") {
    const body: InsertTabletRequest = {
      ...common,
      includesCharger: updates.includesCharger ?? "Unknown",
      workingBattery: updates.workingBattery ?? undefined,
    };
    await apiPutVoid(`/devices/tablet/${id}`, body);
  } else {
    throw new TypeError("Unrecognized device type");
  }

  return apiGet<AnyDevice>(`/devices/${id}`);
}

function chaptersParam(chapterIds: number[]): string {
  return chapterIds.length > 0 ? `?chapters=${chapterIds.join(",")}` : "";
}

function chaptersAndMonthsParam(chapterIds: number[], months: number): string {
  const parts: string[] = [`months=${months}`];
  if (chapterIds.length > 0) parts.push(`chapters=${chapterIds.join(",")}`);
  return `?${parts.join("&")}`;
}

export async function getDashboardCounts(chapterIds: number[]): Promise<DashboardCountsResponse> {
  const params = chapterIds.length > 0 ? `?chapters=${chapterIds.join(",")}` : "";
  return apiGet<DashboardCountsResponse>(`/devices/stats/counts${params}`);
}

export async function getDeviceCount(
  type: "desktop" | "laptop" | "tablet" | "total",
  status: "active" | "not-started" | "in-progress" | "ready-to-donate" | "donated",
  chapterIds: number[]
): Promise<number> {
  return apiGet<number>(
    `/devices/stats/count/${type}?status=${status}${chapterIds.length > 0 ? `&chapters=${chapterIds.join(",")}` : ""}`
  );
}

export async function getAvgTimeInInventory(
  chapterIds: number[]
): Promise<AvgTimeInInventoryResponse> {
  return apiGet<AvgTimeInInventoryResponse>(`/devices/stats/avg-time${chaptersParam(chapterIds)}`);
}

export async function getCompletionRate(chapterIds: number[]): Promise<CompletionRateResponse> {
  return apiGet<CompletionRateResponse>(
    `/devices/stats/completion-rate${chaptersParam(chapterIds)}`
  );
}

export async function getChapterActivityStats(): Promise<ChapterActivityStatsResponse> {
  return apiGet<ChapterActivityStatsResponse>("/devices/stats/chapter-activity");
}

export async function getDevicesReceived(
  chapterIds: number[],
  months = 12
): Promise<MonthlyCountPoint[]> {
  return apiGet<MonthlyCountPoint[]>(
    `/devices/stats/devices-received${chaptersAndMonthsParam(chapterIds, months)}`
  );
}

export async function getDevicesDonated(
  chapterIds: number[],
  months = 12
): Promise<MonthlyCountPoint[]> {
  return apiGet<MonthlyCountPoint[]>(
    `/devices/stats/devices-donated${chaptersAndMonthsParam(chapterIds, months)}`
  );
}

export async function getDonatedDeviceValue(
  chapterIds: number[],
  months = 12
): Promise<MonthlyValuePoint[]> {
  return apiGet<MonthlyValuePoint[]>(
    `/devices/stats/donated-value${chaptersAndMonthsParam(chapterIds, months)}`
  );
}

export async function getDeviceChangelog(id: number): Promise<DeviceChangelogEntry[]> {
  const raw = await apiGet<DeviceChangelogResponse[]>(`/devices/${id}/changelog`);
  return raw.map((e) => ({ ...e, assetId: e.deviceID ?? 0 }));
}

File: backend/src/main/java/org/binaryheart/services/DeviceService.java
package org.binaryheart.services;

import com.google.inject.Inject;
import java.security.InvalidParameterException;
import java.sql.SQLException;
import java.util.List;
import org.binaryheart.exceptions.DeviceNotFoundException;
import org.binaryheart.exceptions.DuplicateKeyException;
import org.binaryheart.exceptions.ForbiddenException;
import org.binaryheart.requests.DeviceListRequest;
import org.binaryheart.repositories.DeviceRepository;
import org.binaryheart.requests.InsertDesktopRequest;
import org.binaryheart.requests.InsertLaptopRequest;
import org.binaryheart.requests.InsertTabletRequest;
import org.binaryheart.responses.AvgTimeInInventoryResponse;
import org.binaryheart.responses.ChapterActivityStatsResponse;
import org.binaryheart.responses.ChapterInventorySummary;
import org.binaryheart.responses.CompletionRateResponse;
import org.binaryheart.responses.DashboardCountsResponse;
import org.binaryheart.responses.DeviceChangelogResponse;
import org.binaryheart.responses.GetDeviceResponse;
import org.binaryheart.responses.MonthlyCountPoint;
import org.binaryheart.responses.MonthlyValuePoint;

public class DeviceService {

	private final DeviceRepository repository;
	private final ChapterService chapterService;

	@Inject
	public DeviceService(DeviceRepository repository, ChapterService chapterService) {
		this.repository = repository;
		this.chapterService = chapterService;
	}

	public int getDeviceCount(String type, String status, List<Integer> requestedChapterIds,
		List<Integer> userChapterIds) throws ForbiddenException, SQLException {
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(requestedChapterIds, userChapterIds);
		return repository.getDeviceCountByChapters(type, status, effectiveChapterIds);
	}

	public DashboardCountsResponse getDashboardCounts(List<Integer> requestedChapterIds, List<Integer> userChapterIds)
		throws ForbiddenException, SQLException {
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(requestedChapterIds, userChapterIds);
		return repository.getDashboardCounts(effectiveChapterIds);
	}

	public AvgTimeInInventoryResponse getAvgTimeInInventory(List<Integer> requestedChapterIds,
		List<Integer> userChapterIds) throws ForbiddenException, SQLException {
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(requestedChapterIds, userChapterIds);
		return repository.getAvgTimeInInventory(effectiveChapterIds);
	}

	public CompletionRateResponse getCompletionRate(List<Integer> requestedChapterIds, List<Integer> userChapterIds)
		throws ForbiddenException, SQLException {
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(requestedChapterIds, userChapterIds);
		return repository.getCompletionRate(effectiveChapterIds);
	}

	public ChapterActivityStatsResponse getChapterActivityStats(List<Integer> userChapterIds)
		throws ForbiddenException, SQLException {
		// Chapter activity stats always use all chapters visible to the user
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(List.<Integer>of(), userChapterIds);
		return repository.getChapterActivityStats(effectiveChapterIds);
	}

	public List<MonthlyCountPoint> getDevicesReceived(List<Integer> requestedChapterIds, List<Integer> userChapterIds,
		int months) throws ForbiddenException, SQLException {
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(requestedChapterIds, userChapterIds);
		return repository.getDevicesReceived(effectiveChapterIds, months);
	}

	public List<MonthlyCountPoint> getDevicesDonated(List<Integer> requestedChapterIds, List<Integer> userChapterIds,
		int months) throws ForbiddenException, SQLException {
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(requestedChapterIds, userChapterIds);
		return repository.getDevicesDonated(effectiveChapterIds, months);
	}

	public List<MonthlyValuePoint> getDonatedDeviceValue(List<Integer> requestedChapterIds,
		List<Integer> userChapterIds, int months) throws ForbiddenException, SQLException {
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(requestedChapterIds, userChapterIds);
		return repository.getDonatedDeviceValue(effectiveChapterIds, months);
	}

	public GetDeviceResponse getDevice(int id) throws DeviceNotFoundException, SQLException {
		return repository.getDevice(id);
	}

	public void deleteDevice(int id, String username) throws DeviceNotFoundException, SQLException {
		try {
			repository.deleteDevice(id, username);
		} catch (SQLException e) {
			if ("02000".equals(e.getSQLState())) {
				throw new DeviceNotFoundException("Could not find device with specified ID: " + id);
			}
			throw e;
		}
	}

	/**
	 * Returns a page of devices scoped to the caller's chapters
	 * ({@code userChapterIds}), optionally narrowed to one {@code chapterId} (the
	 * UI filter, {@code null} for all).
	 */
	public List<GetDeviceResponse> getDevices(List<Integer> userChapterIds, Integer chapterId, DeviceListRequest q)
		throws SQLException, ForbiddenException {
		if (userChapterIds == null || userChapterIds.isEmpty())
			return List.of();
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(chapterId, userChapterIds);
		return repository.getDevices(effectiveChapterIds, q);
	}

	public List<ChapterInventorySummary> getChapterInventorySummary(List<Integer> userChapterIds)
		throws SQLException, ForbiddenException {
		if (userChapterIds == null || userChapterIds.isEmpty())
			return List.of();
		List<Integer> effectiveChapterIds = chapterService.resolveChapterIds(List.of(), userChapterIds);
		return repository.getChapterInventorySummary(effectiveChapterIds);
	}

	public int insertDesktop(InsertDesktopRequest request, String username) throws DuplicateKeyException, SQLException {
		try {
			return repository.insertDesktop(request, username);
		} catch (SQLException e) {
			if ("23505".equals(e.getSQLState())) {
				throw new DuplicateKeyException("An asset with the same asset ID already exists: " + request.assetId());
			} else {
				throw e;
			}
		}
	}

	public int insertLaptop(InsertLaptopRequest request, String username) throws DuplicateKeyException, SQLException {
		try {
			return repository.insertLaptop(request, username);
		} catch (SQLException e) {
			if ("23505".equals(e.getSQLState())) {
				throw new DuplicateKeyException("An asset with the same asset ID already exists: " + request.assetId());
			} else {
				throw e;
			}
		}
	}

	public int insertTablet(InsertTabletRequest request, String username) throws DuplicateKeyException, SQLException {
		try {
			return repository.insertTablet(request, username);
		} catch (SQLException e) {
			if ("23505".equals(e.getSQLState())) {
				throw new DuplicateKeyException("An asset with the same asset ID already exists: " + request.assetId());
			} else {
				throw e;
			}
		}
	}

	public void updateDesktop(InsertDesktopRequest request, String username)
		throws DeviceNotFoundException, SQLException {
		try {
			repository.updateDesktop(request, username);
		} catch (SQLException e) {
			if ("02000".equals(e.getSQLState())) {
				throw new DeviceNotFoundException("Could not find desktop with specified ID: " + request.assetId());
			} else {
				throw e;
			}
		}
	}

	public void updateLaptop(InsertLaptopRequest request, String username)
		throws DeviceNotFoundException, SQLException {
		try {
			repository.updateLaptop(request, username);
		} catch (SQLException e) {
			if ("02000".equals(e.getSQLState())) {
				throw new DeviceNotFoundException("Could not find laptop with specified ID: " + request.assetId());
			} else {
				throw e;
			}
		}
	}

	public void updateTablet(InsertTabletRequest request, String username)
		throws DeviceNotFoundException, SQLException {
		try {
			repository.updateTablet(request, username);
		} catch (SQLException e) {
			if ("02000".equals(e.getSQLState())) {
				throw new DeviceNotFoundException("Could not find tablet with specified ID: " + request.assetId());
			} else {
				throw e;
			}
		}
	}

	public DeviceChangelogResponse[] getDeviceChangelog(List<Integer> userChapterIds, Integer deviceID)
		throws SQLException, InvalidParameterException {
		GetDeviceResponse device = repository.getDevice(deviceID);
		Integer chapterID = chapterService.getChapterIdByName(device.chapter());
		if (device == null || (!userChapterIds.contains(chapterID)
			&& !userChapterIds.contains(chapterService.getNationalChapterId()))) {
			throw new InvalidParameterException("Device not found");
		}

		return repository.getDeviceChangelog(deviceID);
	}
}
